src/sentry/utils/arroyo_router.py

```diff
@@ -0,0 +1,143 @@
+        routing_func: Callable[[Message[TStrategyPayload]], str],
+    ) -> None:
+        self.buffer: MessageBuffer[TResult] = MessageBuffer(list(route_builders.keys()))
+        self.routing_func = routing_func
+        self.next_step = next_step
+
+        self.routes: Mapping[str, ProcessingStrategy[TStrategyPayload]] = {
+            routing_key: builder(MergeRoutes(routing_key, self.buffer))
+            for (routing_key, builder) in route_builders.items()
+        }
+
+    def submit(self, message: Message[TStrategyPayload]) -> None:
+        if len(self.buffer) > 10000:
+            raise MessageRejected
+
+        message_route = self.routing_func(message)
+        self.routes[message_route].submit(message)
+        self.buffer.remove(message, message_route)
+
+        while True:
+            next_message = self.buffer.poll()
+            if next_message:
+                self.next_step.submit(next_message)  # this is fine because of backpresure_guard
+            else:
+                break
+
+    def poll(self) -> None:
+        for route in self.routes.values():
+            route.poll()
+
+        self.next_step.poll()
+
+    def join(self, timeout: float | None = None) -> None:
+        deadline = time.time() + timeout if timeout else None
+        for route in self.routes.values():
+            remaining = deadline - time.time() if deadline else None
+            route.join(remaining)
+
+        remaining = deadline - time.time() if deadline else None
+        self.next_step.join(remaining)
+
+    def close(self) -> None:
+        for route in self.routes.values():
+            route.close()
+        self.next_step.close()
+
+    def terminate(self) -> None:
+        for route in self.routes.values():
+            route.terminate()
+        self.next_step.terminate()
+
+
+class MessageBuffer(Generic[TResult]):
+    """
+    Keeps track of the in-flight offsets for all routes, and buffers messages for submission
+    to the next step so that messages to subsequent stages are submitted in order.
+    """
+
+    def __init__(self, routes: Sequence[str]) -> None:
+        # Maintains the last committed offsets for each route and partition
+        self.committed_offsets: Mapping[str, MutableMapping[Partition, int]] = {
+            r: {} for r in routes
+        }
+        # Keeps track of all messages together with the route on which it was sent
+        self.messages: Deque[tuple[str, Message[TResult]]] = deque()
+
+    def add(self, message: Message[TResult], routing_key: str) -> None:
+        self.messages.append((routing_key, message))
+
+    def remove(self, message: Message[Any], routing_key: str) -> None:
+        for partition, committable_offset in message.committable.items():
+
+            if self.committed_offsets[routing_key].get(partition):
+                if committable_offset > self.committed_offsets[routing_key][partition]:
+                    self.committed_offsets[routing_key][partition] = committable_offset
```

Comment:
Does `self.committed_offsets` refer to actually committed offsets in Kafka? It does not seems from the implementation as we remove a message from the buffer right after submitting.

Reply:
sorry, this was supposed to be `committable_offsets` not `committed_offsets`